src/main/java/org/ice4j/util/AsyncQueueHandler.java

```diff
@@ -0,0 +1,261 @@
+     * asyncronous reading provided queue from thread borrowed from executor to
+     * process items with provided handler.
+     * @param queue thread-safe queue which holds items to process
+     * @param handler an implementation of handler routine which will be
+     * invoked per each item placed in the queue.
+     * @param id optional identifier of current handler for debug purpose
+     * @param executor optional executor service to borrow threads from
+     * @param maxSequentiallyHandledItems maximum number of items sequentially
+     * handled on thread borrowed from {@link #executor} before temporary
+     * releasing thread and re-acquiring it from {@link #executor}.
+     */
+    public AsyncQueueHandler(
+        BlockingQueue<T> queue,
+        Handler<T> handler,
+        String id,
+        ExecutorService executor,
+        long maxSequentiallyHandledItems)
+    {
+        if (queue == null)
+        {
+            throw new IllegalArgumentException("queue is null");
+        }
+
+        if (handler == null)
+        {
+            throw new IllegalArgumentException("handler is null");
+        }
+
+        this.executor = executor != null ? executor : sharedExecutor;
+        this.queue = queue;
+        this.handler = handler;
+        this.id = id;
+        this.maxSequentiallyHandledItems = maxSequentiallyHandledItems;
+    }
+
+    /**
+     * Attempts to stop execution of {@link #reader} if running
+     */
+    public void cancel()
+    {
+        running.set(false);
+        cancel(true);
+    }
+
+    /**
+     * Checks if {@link #reader} is running on one of {@link #executor}
+     * thread and if no submits execution of {@link #reader} on executor.
+     */
+    public void handleQueueItemsUntilEmpty()
+    {
+        synchronized (syncRoot)
+        {
+            if (readerFuture == null || readerFuture.isDone())
+            {
+                running.set(true);
+                readerFuture = executor.submit(reader);
+            }
+        }
+    }
+
+    /**
+     * Invoked when execution of {@link #reader} is about to temporary
+     * cancel and further execution need to be re-scheduled.
+     * Assuming called when lock on {@link #syncRoot} is already taken.
+     */
+    private void onYield()
+    {
+        if (logger.isLoggable(java.util.logging.Level.FINE))
+        {
+            logger.fine("Yielding AsyncQueueHandler associated with "
+                + "AsyncQueueHandler with ID = " + id);
+        }
+
+        synchronized (syncRoot)
+        {
+            cancel(false);
+            readerFuture = executor.submit(reader);
+        }
+    }
+
+    /**
+     * Attempts to cancel currently running reader. Assuming called when
+     * lock on {@link #syncRoot} is already taken
+     * @param mayInterruptIfRunning indicates if {@link #reader} allowed
+     * to be interrupted if running
+     */
+    private void cancel(boolean mayInterruptIfRunning)
+    {
+        synchronized (syncRoot)
```

Comment:
i think the `running.set(false);` needs to be in this method, no?  some inner methods call this `cancel(boolean)` directly, and therefore won't change `running` since it's only set in `cancel()`?

Reply:
> i think the running.set(false); needs to be in this method, no? 

Actually no, because, for example `cancel(boolean)` is called from `onYield`, which just reschedules `reader`, so if I add `running.set(false)` to `cancel(boolean)` then `onYield()` must immediately undo this flag by calling `running.set(true)`...
and it feels like `running` is actually not what it's named, the actual reader may not physically run, but being in queue of corresponding `ExecutorService` waiting it's chance to execute.... So maybe it's better to rollback to negative name which exactly describes what that flag was actually about:  it had meaning that no processing and no further re-scheduling should be done because user of class asked to stop processing?

Reply:
ah i see.  in my opinion it's not `running` that's misleading there.  actually, i wonder if we should extract the onYield code to some `rescheduleReader` private helper? like:
```
// 
private void rescheduleReader()
{
  synchronized (syncRoot)
  {
    if (readerFuture != null)
    {
      readerFuture.cancel(false);
    }
    readerFuture = executor.submit(reader);
  }
}
```

maybe `handleQueueItemsUntilEmpty` could also use this?

Reply:
Done this. It's occurred that after removing extra future used for throttling there no need to cancel existing future in `onYield`, it's possible just override the variable. I've added `rescheduleFuture` which is called from `onYield` and `handleQueueItemsUntilEmpty `. So, now loos better, IMO.